packages/flutter_tools/lib/src/base/process_manager.dart

```diff
@@ -479,3 +491,294 @@ class _RecordingProcess implements Process {
+    Map<String, dynamic> entry = _popEntry(executable, arguments,
+        stdoutEncoding: stdoutEncoding, stderrEncoding: stderrEncoding);
+    return await _ReplayProcessResult.create(_dir, entry);
+  }
+
+  @override
+  ProcessResult runSync(
+      String executable,
+      List<String> arguments,
+      {String workingDirectory,
+       Map<String, String> environment,
+       Encoding stdoutEncoding: SYSTEM_ENCODING,
+       Encoding stderrEncoding: SYSTEM_ENCODING}) {
+    Map<String, dynamic> entry = _popEntry(executable, arguments,
+        stdoutEncoding: stdoutEncoding, stderrEncoding: stderrEncoding);
+    return _ReplayProcessResult.createSync(_dir, entry);
+  }
+
+  /// Finds and returns the next entry in the process manifest that matches
+  /// the specified process arguments. Once found, it marks the manifest entry
+  /// as having been invoked and thus not eligible for invocation again.
+  Map<String, dynamic> _popEntry(String executable, List<String> arguments, {
+    ProcessStartMode mode,
+    Encoding stdoutEncoding,
+    Encoding stderrEncoding,
+  }) {
+    ListsEqual<String> equal = const ListEquality<String>().equals;
+    Map<String, dynamic> entry = _manifest.firstWhere(
+      (Map<String, dynamic> entry) {
+        // Ignore workingDirectory & environment, as they could
+        // yield false negatives.
+        return entry['executable'] == executable
+            && equal(entry['arguments'], arguments)
+            && entry['mode'] == mode?.toString()
+            && entry['stdoutEncoding'] == stdoutEncoding?.name
+            && entry['stderrEncoding'] == stderrEncoding?.name
+            && !(entry['invoked'] ?? false);
+      },
+      orElse: () => null,
+    );
+
+    if (entry == null)
+      throw new StateError('No matching invocation found for $executable');
+
+    entry['invoked'] = true;
+    return entry;
+  }
+
+  @override
+  bool killPid(int pid, [ProcessSignal signal = ProcessSignal.SIGTERM]) {
+    throw new UnimplementedError();
+  }
+}
+
+/// A [ProcessResult] implementation that derives its data from a recording
+/// fragment.
+class _ReplayProcessResult implements ProcessResult {
+  @override
+  final int pid;
+
+  @override
+  final int exitCode;
+
+  @override
+  final dynamic stdout;
+
+  @override
+  final dynamic stderr;
+
+  _ReplayProcessResult._({this.pid, this.exitCode, this.stdout, this.stderr});
+
+  static Future<_ReplayProcessResult> create(
+    Directory dir,
+    Map<String, dynamic> entry,
+  ) async {
+    String basePath = path.join(dir.path, entry['basename']);
+    return new _ReplayProcessResult._(
+      pid: entry['pid'],
+      exitCode: entry['exitCode'],
+      stdout: await _getData('$basePath.stdout', entry['stdoutEncoding']),
+      stderr: await _getData('$basePath.stderr', entry['stderrEncoding']),
+    );
+  }
+
+  static Future<dynamic> _getData(String path, String encoding) async {
+    File file = new File(path);
+    assert(await file.exists());
```

Comment:
I'd kill the assert here and below. This feels more like a legit case for a runtime exception (e.g., `FileSystemException` which you'll get from the read calls below) than a logical invariant violation on the part of the programmer. Filesystems can get unmounted at runtime, etc.